From b6b631e9970b80007b5dc7b1e3d22d185dac1038 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:36:11 -0400
Subject: [ethernet] i40e: don't handle VF reset on unload

Message-id: <1409658042-22689-11-git-send-email-sassmann@redhat.com>
Patchwork-id: 90205
O-Subject: [RHEL7.1 PATCH 010/281] i40e: don't handle VF reset on unload
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

Set the DOWN flag before attempting to disable VFs when unloading the
driver. Also, don't attempt to reset the VFs when the driver is
unloading, because the switch configuration will fail. This fixes a
panic on unload when VFs are enabled.

Change-ID: I25a6567e89c9687145f510ff4f630932412c5c5d
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Signed-off-by: Jesse Brandeburg <jesse.brandeburg@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit eb2d80bcf6b3c101ffe728554ebb702db096917a)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_main.c b/drivers/net/ethernet/intel/i40e/i40e_main.c
index 897452d..628e917 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_main.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_main.c
@@ -8181,16 +8181,16 @@ static void i40e_remove(struct pci_dev *pdev)
 
  i40e_ptp_stop(pf);
 
- if (pf->flags & I40E_FLAG_SRIOV_ENABLED) {
-  i40e_free_vfs(pf);
-  pf->flags &= ~I40E_FLAG_SRIOV_ENABLED;
- }
-
  /* no more scheduling of any task */
  set_bit(__I40E_DOWN, &pf->state);
  del_timer_sync(&pf->service_timer);
  cancel_work_sync(&pf->service_task);
 
+ if (pf->flags & I40E_FLAG_SRIOV_ENABLED) {
+  i40e_free_vfs(pf);
+  pf->flags &= ~I40E_FLAG_SRIOV_ENABLED;
+ }
+
  i40e_fdir_teardown(pf);
 
  /* If there is a switch structure or any orphans, remove them.
diff --git a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
index 9074f63..7d133fa 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
@@ -1868,7 +1868,8 @@ int i40e_vc_process_vflr_event(struct i40e_pf *pf)
    /* clear the bit in GLGEN_VFLRSTAT */
    wr32(hw, I40E_GLGEN_VFLRSTAT(reg_idx), (1 << bit_idx));
 
-   i40e_reset_vf(vf, true);
+   if (!test_bit(__I40E_DOWN, &pf->state))
+    i40e_reset_vf(vf, true);
   }
  }
 
-- 
1.7.1